### PR TITLE
Report pre-warmed app starts

### PR DESCRIPTION
**Pull request by philipphofmann**
## :scroll: Description

Report pre-warmed app starts by dropping the first app start spans if pre-warming paused during these steps.
This approach will shorten the app start duration, but it represents the duration a user has to wait after clicking
the app icon until the app is responsive. We report the app start type in the appContext, so Sentry can make
changes to the UI for prewarmed app starts.

As it's really hard to test this properly, I added an experimental feature flag so we can get feedback first before making it GA. I plan to remove the feature flag in the future if we are sure this works correctly.

Docs PR https://github.com/getsentry/sentry-docs/pull/5762

Normal app start with all spans

image

Pre-warmed with dropped spans

image

App Context
image

## :bulb: Motivation and Context

Fixes GH-1897

## :green_heart: How did you test it?
Unit tests and simulators

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [x] I updated the docs if needed
- [x] Review from the native team if needed
- [x] No breaking changes

### CODE REVIEW

**Comment by philipphofmann** [comment]
@mitsuhiko, @k-fish, and I agreed on reporting the app start type in the app context as  `start_type : cold | cold.prewarmed | warm | warm.prewarmed`. [Internal Notion Page](https://www.notion.so/sentry/Cocoa-Report-Prewarmed-App-Starts-2a48da2973364728b7ee769daee34064#8109159ee5bd411c9c281a26e1c95607) for more context.

**Review by brustolin — COMMENTED** [review]
First look. Just one thing that I believe need to change.

File: Sources/Sentry/NSMutableDictionary+Sentry.m
Comment (philipphofmann) [reply Sources/Sentry/NSMutableDictionary+Sentry.m]: Excellent catch. Totally missed that thanks 🥇

**Review by brustolin — APPROVED** [review]
LGTM

**Comment by philipphofmann** [comment]
Please don't merge yet, waiting for a green light that pre-warmed app starts are searchable in discover.

**Comment by philipphofmann** [comment]
Blocked by https://github.com/getsentry/sentry/issues/36927

File: Sources/Sentry/SentryAppStartTracker.m
Comment (armcknight) [inline Sources/Sentry/SentryAppStartTracker.m:152]: m: I think that 

> Prewarming executes an app’s launch sequence up until, but not including, when main() calls UIApplicationMain 

is consistent with 

> Our findings show that most of the prewarmed app starts don't call the main method

I'd just state our summary of the situation as we've observed it:

```suggestion
        // Prewarming can stop at any of the app launch steps. Our findings show that most of
        // the prewarmed app starts don't call the main method. Therefore we subtract the 
        // time before the module initialization / main method to calculate the app start 
        // duration. If the app start stopped during a later launch step, we drop it below with 
        // checking the SENTRY_APP_START_MAX_DURATION. With this approach, we will 
        // lose some warm app starts, but we accept this tradeoff. Useful resources:
```

File: Sources/Sentry/SentryTracer.m
Comment (armcknight) [inline Sources/Sentry/SentryTracer.m:646]: l: the idiomatic way to create a new array is:
```suggestion
    NSMutableArray<SentrySpan *> *appStartSpans = [NSMutableArray array];
```

File: Tests/SentryTests/Categories/NSMutableDictionarySentryTests.swift
Comment (armcknight) [inline Tests/SentryTests/Categories/NSMutableDictionarySentryTests.swift:53]: What are the real-world equivalents of these? Seems like it'd lead to some kind of data loss. Or is it upgrading the type of something? Seems like we'd more want to a merge produce an array of values for a particular key if it already exists, but that leads to an issue of heterogenous types in that array. Do we have such heterogeneity anywhere in practice?

File: Tests/SentryTests/Helper/TestSysctl.swift
Comment (armcknight) [inline Tests/SentryTests/Helper/TestSysctl.swift:25]: m: this would always create a date of January 1 at time midnight UTC. Are we using this as a sentinel value? Why not `nil`?

File: Tests/SentryTests/SentryScopeSwiftTests.swift
Comment (armcknight) [inline Tests/SentryTests/SentryScopeSwiftTests.swift:265]: m: if we decide to keep the current overwrite logic for unequal types, we should also write a test case for that here.

**Review by armcknight — CHANGES_REQUESTED** [review]
Couple questions for ya!

File: Sources/Sentry/SentryTracer.m
Comment (brustolin) [reply Sources/Sentry/SentryTracer.m:646]: Any practical differences? Otherwise, I suggest sticking with the first approach, for me it's friendlier as it's the same for all classes.

File: Tests/SentryTests/Categories/NSMutableDictionarySentryTests.swift
Comment (brustolin) [reply Tests/SentryTests/Categories/NSMutableDictionarySentryTests.swift:53]: I think @philipphofmann meant to make a more generic test, objc Dictionary it is not strongly typed and anything can happen. 
Since the code this is testing is an auxiliary non specific code, it's worth testing this behaviour.

File: Tests/SentryTests/Helper/TestSysctl.swift
Comment (brustolin) [reply Tests/SentryTests/Helper/TestSysctl.swift:25]: It's just a fixed value we can safely increment and predict the outcome.

File: Tests/SentryTests/SentryScopeSwiftTests.swift
Comment (brustolin) [reply Tests/SentryTests/SentryScopeSwiftTests.swift:265]: Quick question since Im no specialist in tests. If the other test succeeds and keeps the behaviour at check, wouldn't be just a repetition to have the same test here? Here we just need to make sure both `context`  are being merged, or Am I missing something?

File: Sources/Sentry/SentryTracer.m
Comment (armcknight) [reply Sources/Sentry/SentryTracer.m:646]: I built a habit of not using `+[new]` in pre-ARC code, but even with ARC now I don't use it as it can lead to weirdness with subclasses not calling designated initializers, like `NSManagedObject`. Whether or not `new` is used, we all need to be aware of the designated init for a given class, so it doesn't matter much what we end up committing, I guess. I just like not having the extra indirection, or having to double check every time I see a `new` whether or not `init` is OK to use for that class. `new` can also hide issues introduced in later API: https://blog.cristik.com/2018/02/objective-c-new-is-dangerous-and-old-avoid-it-at-all-costs/

File: Tests/SentryTests/Categories/NSMutableDictionarySentryTests.swift
Comment (armcknight) [reply Tests/SentryTests/Categories/NSMutableDictionarySentryTests.swift:53]: It's being used to merge contexts here: https://github.com/getsentry/sentry-cocoa/pull/1969/files#diff-8ba9297eda94f834f1fc0767e1550efd533b9a402b79f85bf829f64a0f04b7b0L515-L521 

If this can drop data from someone's context, I think it would be a better approach to at least check the types inside the dictionary during the merge and print a warning to let people know they're going to lose data in the merge due to heterogenous types.

The reason I asked about real-world equivalents is because I'd think ideally we'd move away from dictionaries altogether, especially if its something that's going to be parsed by our backend. But if this is something that only customers will ever write to and read from, then yes, they should have the freedom to do what they want.

File: Tests/SentryTests/SentryScopeSwiftTests.swift
Comment (armcknight) [reply Tests/SentryTests/SentryScopeSwiftTests.swift:265]: I'm thinking that if someone comes in later and doesn't know that data can be dropped from the scope dictionaries this way, they might build on that assumption which could break other functionality in an unexpected way, either at the time they introduce their change, or when a change comes in later that also uses that behavior. It would be helpful to have a test that can fail in that specific case.

File: Sources/Sentry/SentryTracer.m
Comment (philipphofmann) [reply Sources/Sentry/SentryTracer.m:647]: Didn't know that, thanks.

Comment (philipphofmann) [reply Sources/Sentry/SentryTracer.m:647]: I created an issue to create a linter for that https://github.com/getsentry/sentry-cocoa/issues/2126

File: Tests/SentryTests/Categories/NSMutableDictionarySentryTests.swift
Comment (philipphofmann) [reply Tests/SentryTests/Categories/NSMutableDictionarySentryTests.swift:53]: We use dictionaries for the context, so users can put in their custom data if they want to. Dictionaries give them the advantage of using deep nesting for their data. We add the app start metrics to that context. The problem is that we need to update a nested value inside that dictionary. I just got tired of writing complex code of getting the dictionary inside a key of the context dictionary, making it mutable, adding a new value, and then adding it again to the dictionary. Therefore, I added `mergeEntries` to make this process easier.

We could start a discussion if we want to switch to a typed context, which still allows you to add your custom nested data, but this would be a breaking change, and we have to wait until the next major. 

Most Sentry APIs are designed for languages where dictionaries are standard, like JavaScript and Python. Sometimes we need to change them and adapt them to strongly typed languages, Swift and ObjC.

File: Tests/SentryTests/SentryScopeSwiftTests.swift
Comment (philipphofmann) [reply Tests/SentryTests/SentryScopeSwiftTests.swift:265]: Done

File: Sources/Sentry/SentryClient.m
Comment (armcknight) [inline Sources/Sentry/SentryClient.m:658]: This might be an iOS only feature. Do you know if it also exists on tvOS?

File: Sources/Sentry/SentryTracer.m
Comment (armcknight) [inline Sources/Sentry/SentryTracer.m:630]: 🙏🏻

File: Tests/SentryTests/Performance/SentryTracerTests.swift
Comment (armcknight) [inline Tests/SentryTests/Performance/SentryTracerTests.swift:899]: If this `500` is meant to test exceeding the maximum launch duration, we could convert that macro to a constant and expose it to this test.

File: Tests/SentryTests/SentryScopeSwiftTests.swift
Comment (armcknight) [inline Tests/SentryTests/SentryScopeSwiftTests.swift:256]: This is great, thank you!

**Review by armcknight — APPROVED** [review]
Looks good. Would you mind writing some comments to explain some of the constant int values in some of the tests, like [here](https://github.com/getsentry/sentry-cocoa/pull/1969/files/9f5c23a6968327c9176eb8e22ef8f1625ea2a7ed..a248f3b52168ea5edf72fa9a7bd7719b30c3cff6#diff-6dd6077222e81d2de523aea25d4f322960ca9a04690a55346c6f0ecb0cac28c3R562) and [here](https://github.com/getsentry/sentry-cocoa/pull/1969/files/9f5c23a6968327c9176eb8e22ef8f1625ea2a7ed..a248f3b52168ea5edf72fa9a7bd7719b30c3cff6#diff-6dd6077222e81d2de523aea25d4f322960ca9a04690a55346c6f0ecb0cac28c3R576)?

**Review by brustolin — APPROVED** [review]
LGTM

File: Sources/Sentry/SentryClient.m
Comment (philipphofmann) [reply Sources/Sentry/SentryClient.m:658]: I'm unsure. As this is an experimental feature, I would like to see if we get any data from tvOS projects to find out.

File: Tests/SentryTests/Performance/SentryTracerTests.swift
Comment (philipphofmann) [reply Tests/SentryTests/Performance/SentryTracerTests.swift:899]: It depends on `fixture.appStartDuration`. Good point, I changed it.

**Pull request merged**
